### ISSUE TITLE
In macros, expr followed by { is buggy

### ISSUE THREAD

**Issue by Stebalien** [rust-lang/rust#26739]
The following compiles:
```rust
macro_rules! test {
    ($e:expr {}) => {$e};
}
```

However, (1) this technically isn't allowed by the macro future proofing rules and (2) it doesn't work when `$e` is an ident because rust tries to parse `ident {}` as an invalid struct literal instantiation.

The following works:

```rust
test! {
    1 + 1 { } // 1 + 1 can be anything other than an ident.
};
```
And the following doesn't:
```
test! {
        ident { }
};
```

Error:

```text
tmp.rs:10:15: 10:16 error: structure literal must either have at least one field or use functional structure update syntax
tmp.rs:10         ident { }
                        ^
tmp.rs:10:17: 10:18 error: unexpected end of macro invocation
tmp.rs:10         ident { }
                          ^
```

**Comment by nikomatsakis** [comment]
triage: P-high (decision needed)

For this and other macro rules, it's unclear how much we can change here, but I think we should make an effort to fix these bugs sooner rather than later if we are going to do so!

**Comment by pnkfelix** [comment]
This was fixed by #30694

**Comment by pnkfelix** [comment]
(though i will note yet again that there may be motivation for adding a new syntactic class to macro rules, which I have been [referring to tongue-in-cheek](https://github.com/rust-lang/rfcs/pull/1384#issuecomment-160165525) as `$e:moelarry`, but obviously we'd want to bikeshed a better name before we'd actually put it in for real...)
